File: 2015/Go/9/part1/main.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

var minDist = math.MaxInt64
var distances = map[string]map[string]int{}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	for scanner.Scan() {
		input := scanner.Text()
		parts := strings.Fields(input)
		cityA := parts[0]
		cityB := parts[2]
		distanceNumber, _ := strconv.Atoi(parts[4])

		if distances[cityA] == nil {
			distances[cityA] = make(map[string]int)
		}
		distances[cityA][cityB] = distanceNumber

		if distances[cityB] == nil {
			distances[cityB] = make(map[string]int)
		}
		distances[cityB][cityA] = distanceNumber
	}
	totalCities := len(distances)

	for startCity := range distances {
		visited := map[string]bool{}
		visited[startCity] = true

		citiesVisited(startCity, visited, 0, totalCities)
	}
	fmt.Println(minDist)
}

func citiesVisited(city string, visited map[string]bool, totalDistance int, totalCities int) {
	if len(visited) == totalCities {
		if totalDistance < minDist {
			minDist = totalDistance
		}
		return
	}
	for neighbor, dist := range distances[city] {
		if !visited[neighbor] {
			visited[neighbor] = true
			citiesVisited(neighbor, visited, totalDistance+dist, totalCities)
			delete(visited, neighbor)
		}
	}

}

File: 2015/Rust/9/part1/src/main.rs
use std::collections::{HashMap, HashSet};

fn main() {
    let raw_input = include_str!("input.txt");
    let mut distances: HashMap<String, HashMap<String, i32>> = HashMap::new();
    for line in raw_input.lines() {
        let parts: Vec<&str> = line.split_whitespace().collect();
        let city_a = parts[0];
        let city_b = parts[2];
        let distance_number = parts[4].parse::<i32>().unwrap_or(0);

        distances
            .entry(city_a.to_string())
            .or_insert_with(HashMap::new)
            .insert(city_b.to_string(), distance_number);

        distances
            .entry(city_b.to_string())
            .or_insert_with(HashMap::new)
            .insert(city_a.to_string(), distance_number);
    }

    let mut min_dist = i32::MAX;
    let total_cities = distances.len();
    for start_city in distances.keys() {
        let mut visited = HashSet::new();
        visited.insert(start_city.clone());

        cities_visited(
            start_city,
            &mut visited,
            0,
            total_cities,
            &distances,
            &mut min_dist,
        );
    }
    println!("{}", min_dist)
}

fn cities_visited(
    city: &str,
    visited: &mut HashSet<String>,
    total_distance: i32,
    total_cities: usize,
    distances: &HashMap<String, HashMap<String, i32>>,
    min_dist: &mut i32,
) {
    if visited.len() == total_cities {
        if total_distance < *min_dist {
            *min_dist = total_distance
        }
        return;
    }

    if let Some(neighbors) = distances.get(city) {
        for (neighbor, &dist) in neighbors {
            if !visited.contains(neighbor) {
                visited.insert(neighbor.clone());

                cities_visited(
                    neighbor,
                    visited,
                    total_distance + dist,
                    total_cities,
                    distances,
                    min_dist,
                );

                visited.remove(neighbor);
            }
        }
    }
}
